File: September 2026/2026-09-06/581-shortest-unsorted-conti-subarray.py
# 581-shortest-unsorted-conti-subarray.cpp;2p;self;leetcode

class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        left = 0
        while left < n - 1 and nums[left] <= nums[left + 1]:
            left += 1

        if left == n - 1:
            return 0

        right = n - 1
        while right > 0 and nums[right - 1] <= nums[right]:
            right -= 1

        sub_min = min(nums[left:right + 1])
        sub_max = max(nums[left:right + 1])

        while left > 0 and nums[left - 1] > sub_min:
            left -= 1

        while right < n - 1 and nums[right + 1] < sub_max:
            right += 1

        return right - left + 1

File: September/2026-09-06/581/581-shortest-unsorted-conti-subarray.cpp
// 581-shortest-unsorted-conti-subarray.cpp;2p;self;leetcode

class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        int n = nums.size();
        int left = 0;

        while (left < n - 1 && nums[left] <= nums[left + 1]) {
            left++;
        }
        if (left == n - 1) {
            return 0;
        }

        int right = n - 1;
        while (right > 0 && nums[right - 1] <= nums[right]) {
            right--;
        }

        int subMin = nums[left];
        int subMax = nums[left];

        for (int i = left; i <= right; i++) {
            subMin = min(subMin, nums[i]);
            subMax = max(subMax, nums[i]);
        }
        while (left > 0 && nums[left - 1] > subMin) {
            left--;
        }
        while (right < n - 1 && nums[right + 1] < subMax) {
            right++;
        }

        return right - left + 1;
    }
};